go/vt/vttablet/tabletserver/query_executor.go

```diff
@@ -387,14 +390,31 @@ func (qre *QueryExecutor) checkPermissions() error {
 		remoteAddr = ci.RemoteAddr()
 		username = ci.Username()
 	}
-	action, desc := qre.plan.Rules.GetAction(remoteAddr, username, qre.bindVars, qre.marginComments)
-	switch action {
-	case rules.QRFail:
-		return vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "disallowed due to rule: %s", desc)
-	case rules.QRFailRetry:
-		return vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "disallowed due to rule: %s", desc)
-	}
 
+	ctx, cancel := context.WithTimeout(qre.ctx, maxQueryBufferDuration)
+	defer cancel()
+	for goodToProceed := false; !goodToProceed; {
+		action, desc := qre.plan.Rules.GetAction(remoteAddr, username, qre.bindVars, qre.marginComments)
+		switch action {
+		case rules.QRFail:
+			return vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "disallowed due to rule: %s", desc)
+		case rules.QRFailRetry:
+			return vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "disallowed due to rule: %s", desc)
+		case rules.QRBuffer:
+			// We buffer up to some timeout. The timeout is determined by ctx.Done().
+			// If we're not at timeout yet, then we sleep for a short amount of time before
+			// resuming the loop.
+			select {
+			case <-time.After(time.Second):
+				// keep iteration
```

Comment:
There will be potentially dozens, or hundreds, or thousands of waiting queries (up to max concurrent connections we allow in vtgate). So they will all be running this busy wait; this is why I picked a relatively relaxed interval.
we can expect cut-over time to be ~2s-3s.
WDYT?

Reply:
Personally, I think 1s is good. There will be at most one extra second to wait for the query to be executed. I think it is a good tradeoff between the query response time and the stress we put on VTGate.

Reply:
I think you are referring to VTTablet. which means the query is already sent down to the tablet. Just thinking in terms of reducing response time on the cutover is complete. This will also allow VTGate to send more queries down to Tablet which were blocked as the earlier query has not responded.

Reply:
Right, but it comes with a cost: an increased busy wait on `tabletserver`, with potentially hundreds or thousands connections asking it the same question: "are we good to go?". So there has to be some tradeoff. I don't know what the exact number is; but if cut-over is expected to take some 3 seconds, then a 1sec wait interval isn't that excessive IMO. I'm good to lower this, say 500ms? We can iterate on that later.

Reply:
For this, I usually see how shard_buffer does it without using timeout. It would be nice we can have something similar. Otherwise the code looks good to me.